File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/repository/query/PartTreeRequeryQuery.java
package com.coupang.springframework.data.requery.repository.query;

import com.coupang.springframework.data.requery.core.RequeryOperations;
import com.coupang.springframework.data.requery.mapping.RequeryMappingContext;
import com.coupang.springframework.data.requery.provider.RequeryPersistenceProvider;
import com.coupang.springframework.data.requery.repository.query.RequeryQueryExecution.DeleteExecution;
import com.coupang.springframework.data.requery.repository.query.RequeryQueryExecution.ExistsExecution;
import io.requery.query.NamedExpression;
import io.requery.query.Scalar;
import io.requery.query.element.QueryElement;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.domain.Sort;
import org.springframework.data.repository.query.ResultProcessor;
import org.springframework.data.repository.query.ReturnedType;
import org.springframework.data.repository.query.parser.PartTree;
import org.springframework.util.Assert;

import static com.coupang.springframework.data.requery.utils.RequeryUtils.*;

/**
 * {@link PartTree} 정보를 바탕으로 Requery {@link QueryElement} 를 빌드합니다.
 *
 * @author debop
 * @since 18. 6. 9
 */
@Slf4j
public class PartTreeRequeryQuery extends AbstractRequeryQuery {

    private final PartTree tree;
    private final RequeryParameters parameters;

    private final QueryPreparer queryPreparer;
    private final CountQueryPreparer countQueryPreparer;
    private final RequeryOperations operations;
    private final RequeryMappingContext context;


    public PartTreeRequeryQuery(@NotNull RequeryQueryMethod method,
                                @NotNull RequeryOperations operations,
                                @NotNull RequeryPersistenceProvider persistenceProvider) {
        super(method, operations);

        Assert.notNull(persistenceProvider, "PersistenceProvider must not be null");

        this.operations = operations;
        this.context = operations.getMappingContext();

        Class<?> domainClass = method.getEntityInformation().getJavaType();
        this.parameters = method.getParameters();

        log.debug("Create PartTreeRequeryQuery. domainClass={}, parameters={}", domainClass, parameters);

        try {
            this.tree = new PartTree(method.getName(), domainClass);
            this.countQueryPreparer = new CountQueryPreparer(persistenceProvider);
            this.queryPreparer = tree.isCountProjection() ? countQueryPreparer : new QueryPreparer(persistenceProvider);
        } catch (Exception e) {
            throw new IllegalArgumentException("Fail to create query for method [" + method + "] message=" + e.getMessage(), e);
        }
    }


    @Override
    protected QueryElement<?> doCreateQuery(Object[] values) {
        return queryPreparer.createQuery(values);
    }

    @SuppressWarnings("unchecked")
    @Override
    protected QueryElement<? extends Scalar<Integer>> doCreateCountQuery(Object[] values) {
        return (QueryElement<? extends Scalar<Integer>>) countQueryPreparer.createQuery(values);
    }

    @Override
    protected RequeryQueryExecution getExecution() {
        if (tree.isDelete()) {
            log.debug("Create DeleteExecution. queryMethod={}", queryMethod);
            return new DeleteExecution(operations);
        } else if (tree.isExistsProjection()) {
            log.debug("Create ExistsExecution. queryMethod={}", queryMethod);
            return new ExistsExecution();
        }
        return super.getExecution();
    }

    @SuppressWarnings("unchecked")
    protected QueryElement<?> prepareQuery(RequeryParameterAccessor accessor) {
        QueryElement<?> query = unwrap(getOperations().select(getDomainClass()));

        query = buildWhereClause(query, accessor);

        if (accessor.getParameters().hasPageableParameter()) {
            query = applyPageable(getDomainClass(), query, accessor.getPageable());
        }
        if (accessor.getParameters().hasSortParameter()) {
            query = applySort(getDomainClass(), query, accessor.getSort());
        }

        return query;
    }

    @SuppressWarnings("unchecked")
    protected QueryElement<?> buildWhereClause(QueryElement<?> baseQuery,
                                               RequeryParameterAccessor accessor) {

        final RequeryParameters bindableParams = accessor.getParameters().getBindableParameters();
        final int bindableParamsSize = bindableParams.getNumberOfParameters();

        QueryElement<?>[] queries = new QueryElement[1];

        queries[0] = baseQuery;
        for (int i = 0; i < bindableParamsSize; i++) {
            final RequeryParameters.RequeryParameter param = bindableParams.getParameter(i);
            final Object value = accessor.getBindableValue(i);

            if (param.isNamedParameter() && param.getName().isPresent()) {
                NamedExpression expr = NamedExpression.of(param.getName().get(), value.getClass());
                queries[0] = unwrap(baseQuery.where(expr.eq(value)));
            }
        }
        return queries[0];
    }


    /**
     * Query preparer to create {@link QueryElement} instances and potentially cache them.
     */
    private class QueryPreparer {

        private final RequeryPersistenceProvider persistenceProvider;

        QueryPreparer(@NotNull RequeryPersistenceProvider persistenceProvider) {
            Assert.notNull(persistenceProvider, "persistenceProvider must not be null!");
            this.persistenceProvider = persistenceProvider;

            // HINT: check wrong method (parameter number matching, not exists property name ...)
            RequeryQueryCreator creator = createCreator(persistenceProvider, null);
            if (creator != null) {
                creator.createQuery();
                creator.getParameterExpressions();
            }
        }

        public QueryElement<?> createQuery(Object[] values) {

            RequeryParametersParameterAccessor accessor = new RequeryParametersParameterAccessor(parameters, values);
            RequeryQueryCreator creator = createCreator(persistenceProvider, accessor);

            QueryElement<?> query = creator.createQuery(getDynamicSort(values));

            return restrictMaxResultsIfNecessary(query);
        }

        private QueryElement<?> restrictMaxResultsIfNecessary(QueryElement<?> baseQuery) {

            QueryElement<?> query = baseQuery;
            if (tree.isLimiting()) {
                if (query.getLimit() != null && query.getLimit() > 0) {
                    /*
                     * In order to return the correct results, we have to adjust the first result offset to be returned if:
                     * - a Pageable parameter is present
                     * - AND the requested page number > 0
                     * - AND the requested page size was bigger than the derived result limitation via the First/Top keyword.
                     */
                    if (query.getLimit() > tree.getMaxResults() && query.getOffset() > 0) {
                        query = unwrap(query.offset(query.getOffset() - (query.getLimit() - tree.getMaxResults())));
                    }
                }
                query = unwrap(query.limit(tree.getMaxResults()));
            }

            if (tree.isExistsProjection()) {
                query = unwrap(query.limit(1));
            }

            return query;
        }

        protected RequeryQueryCreator createCreator(final RequeryPersistenceProvider persistenceProvider,
                                                    final RequeryParametersParameterAccessor accessor) {

            ParameterMetadataProvider provider = (accessor != null)
                                                 ? new ParameterMetadataProvider(accessor, persistenceProvider)
                                                 : new ParameterMetadataProvider(parameters, persistenceProvider);

            ResultProcessor processor = getQueryMethod().getResultProcessor();
            ReturnedType returnedType = (accessor != null)
                                        ? processor.withDynamicProjection(accessor).getReturnedType()
                                        : processor.getReturnedType();

            return new RequeryQueryCreator(operations, provider, returnedType, tree);
        }

        private Sort getDynamicSort(Object[] values) {
            return parameters.potentiallySortsDynamically()
                   ? new RequeryParametersParameterAccessor(parameters, values).getSort()
                   : Sort.unsorted();
        }

    }

    /**
     * Special {@link QueryPreparer} to create count queries.
     */
    private class CountQueryPreparer extends QueryPreparer {

        CountQueryPreparer(@NotNull RequeryPersistenceProvider persistenceProvider) {
            super(persistenceProvider);
        }

        @Override
        protected RequeryCountQueryCreator createCreator(RequeryPersistenceProvider persistenceProvider,
                                                         RequeryParametersParameterAccessor accessor) {
            ParameterMetadataProvider provider = (accessor != null)
                                                 ? new ParameterMetadataProvider(accessor, persistenceProvider)
                                                 : new ParameterMetadataProvider(parameters, persistenceProvider);

            return new RequeryCountQueryCreator(operations,
                                                provider,
                                                getQueryMethod().getResultProcessor().getReturnedType(),
                                                tree);
        }
    }
}

File: spring-data-requery/src/main/kotlin/org/springframework/data/requery/repository/query/PartTreeRequeryQuery.kt
package org.springframework.data.requery.repository.query

import io.requery.query.NamedExpression
import io.requery.query.Result
import io.requery.query.element.QueryElement
import mu.KotlinLogging
import org.springframework.data.domain.Sort
import org.springframework.data.repository.query.parser.PartTree
import org.springframework.data.requery.applyPageable
import org.springframework.data.requery.applySort
import org.springframework.data.requery.core.RequeryOperations
import org.springframework.data.requery.mapping.RequeryMappingContext
import org.springframework.data.requery.provider.RequeryPersistenceProvider
import org.springframework.data.requery.unwrap

/**
 * [PartTree] 정보를 바탕으로 Requery [QueryElement]를 빌드합니다.
 *
 * @author debop@coupang.com
 */
open class PartTreeRequeryQuery(queryMethod: RequeryQueryMethod,
                                operations: RequeryOperations,
                                persistenceProvider: RequeryPersistenceProvider): AbstractRequeryQuery(queryMethod, operations) {

    companion object {
        private val log = KotlinLogging.logger { }
    }

    private val tree: PartTree
    private val parameters: RequeryParameters

    private val countQueryPreparer: CountQueryPreparer
    private val queryPreparer: QueryPreparer
    private val context: RequeryMappingContext

    init {

        context = operations.mappingContext
        parameters = queryMethod.parameters

        log.debug { "Create PartTreeRequeryQuery. domainClass=$domainClass, parameters=$parameters" }

        try {
            this.tree = PartTree(queryMethod.name, domainClass)
            this.countQueryPreparer = CountQueryPreparer(persistenceProvider)
            this.queryPreparer = when {
                tree.isCountProjection -> countQueryPreparer
                else -> QueryPreparer(persistenceProvider)
            }
        } catch(e: Exception) {
            throw IllegalArgumentException("Fail to create query for method [$queryMethod], message=${e.message}", e)
        }
    }

    override fun doCreateQuery(values: Array<Any?>): QueryElement<out Any> {
        return queryPreparer.createQuery(values).unwrap()
    }

    @Suppress("UNCHECKED_CAST")
    override fun doCreateCountQuery(values: Array<Any?>): QueryElement<out Result<Int>> {
        return countQueryPreparer.createQuery(values).unwrap() as QueryElement<out Result<Int>>
    }

    override fun getExecution(): RequeryQueryExecution {
        return when {
            tree.isDelete -> DeleteExecution(operations)
            tree.isExistsProjection -> ExistsExecution()
            else -> super.getExecution()
        }
    }

    protected fun prepareQuery(accessor: RequeryParameterAccessor): QueryElement<out Any> {

        val query = buildWhereClause(operations.select(domainClass).unwrap(), accessor)

        return when {

            accessor.parameters.hasPageableParameter() ->
                query.applyPageable(domainClass, accessor.pageable)

            accessor.parameters.hasSortParameter() ->
                query.applySort(domainClass, accessor.sort)

            else -> query
        }
    }

    protected fun buildWhereClause(baseQuery: QueryElement<out Any>, accessor: RequeryParameterAccessor): QueryElement<out Any> {

        val bindableParams = accessor.parameters.bindableParameters
        var query = baseQuery

        bindableParams.forEachIndexed { index, parameter ->
            val value = accessor.getBindableValue(index)

            if(parameter.isNamedParameter && parameter.name.isPresent) {
                log.trace { "Add named expression ... name=${parameter.name.get()}, value=$value" }
                val expr = NamedExpression.of(parameter.name.get(), value.javaClass)
                query = query.where(expr.eq(value)).unwrap()
            }
        }

        return query
    }

    @Suppress("LeakingThis")
    open inner class QueryPreparer(val provider: RequeryPersistenceProvider) {

        init {
            // HINT: check wrong method (parameter number matching, not exists property name ...)
            createCreator(null)?.let {
                it.createQuery()
                it.parameterExpressions
            }
        }

        fun createQuery(values: Array<Any?>): QueryElement<out Any> {
            val accessor = RequeryParametersParameterAccessor(parameters, values)
            val creator = createCreator(accessor)!!

            var query = creator.createQuery(getDynamicSort(values))

            if(queryMethod.isPageQuery) {
                query = query.applyPageable(domainClass, accessor.pageable)
            }
            return restrictMaxResultsIfNecessary(query)
        }

        private fun restrictMaxResultsIfNecessary(baseQuery: QueryElement<out Any>): QueryElement<out Any> {

            var query = baseQuery

            if(tree.isLimiting) {
                if(query.limit != null) {
                    if(query.offset == null) {
                        query.offset(0)
                    }
                    /*
                     * In order to return the correct results, we have to adjust the first result offset to be returned if:
                     * - a Pageable parameter is present
                     * - AND the requested page number > 0
                     * - AND the requested page size was bigger than the derived result limitation via the First/Top keyword.
                     */
                    if(query.limit > tree.maxResults ?: 0 && query.offset > 0) {
                        val offset = query.offset - (query.limit - tree.maxResults!!)
                        query = query.offset(offset).unwrap()
                    }
                }
                query = query.limit(tree.maxResults!!).unwrap()
            }

            if(tree.isExistsProjection) {
                query = query.limit(1).unwrap()
            }

            return query
        }

        protected open fun createCreator(accessor: RequeryParametersParameterAccessor?): RequeryQueryCreator? {

            val provider = when(accessor) {
                null -> ParameterMetadataProvider(provider, parameters = parameters)
                else -> ParameterMetadataProvider(provider, accessor = accessor)
            }

            val processor = queryMethod.resultProcessor

            val returnedType = when(accessor) {
                null -> processor.returnedType
                else -> processor.withDynamicProjection(accessor).returnedType
            }

            return RequeryQueryCreator(operations, provider, returnedType, tree)

        }

        private fun getDynamicSort(values: Array<Any?>): Sort = when {
            parameters.potentiallySortsDynamically() -> RequeryParametersParameterAccessor(parameters, values).sort
            else -> Sort.unsorted()
        }
    }

    /**
     * Special [QueryPreparer] to create count queries.
     */
    open inner class CountQueryPreparer(provider: RequeryPersistenceProvider): QueryPreparer(provider) {

        override fun createCreator(accessor: RequeryParametersParameterAccessor?): RequeryQueryCreator? {

            val provider = when(accessor) {
                null -> ParameterMetadataProvider(provider, parameters = parameters)
                else -> ParameterMetadataProvider(provider, accessor = accessor)
            }

            return RequeryCountQueryCreator(operations,
                                            provider,
                                            queryMethod.resultProcessor.returnedType,
                                            tree)
        }
    }
}